src/item.cpp

```diff
@@ -4334,6 +4190,110 @@ bool item::gunmod_compatible( const item& mod, bool alert, bool effects ) const
     return false;
 }
 
+std::map<std::string, const item::gun_mode> item::gun_all_modes() const
+{
+    if( !is_gun() ) {
+        return {};
+    }
+
+    std::map<std::string, const item::gun_mode> res;
+
+    auto opts = gunmods();
+    opts.push_back( this );
+
+    for( auto e : opts ) {
+        if( e->is_gun() ) {
+            for( auto m : e->type->gun->modes ) {
+                // prefix attached gunmods, eg. M203_DEFAULT to avoid index key collisions
+                std::string prefix = e->is_gunmod() ? ( std::string( e->typeId() ) += "_" ) : "";
+                std::transform( prefix.begin(), prefix.end(), prefix.begin(), (int(*)(int))std::toupper );
+
+                auto qty = m.second.second;
+                if( m.first == "AUTO" && e == this && has_flag( "RAPIDFIRE" ) ) {
+                    qty *= 1.5;
+                }
+
+                res.emplace( prefix += m.first, item::gun_mode { m.second.first, const_cast<item *>( e ),
+                                                                 qty, false } );
+            };
+        }
+        if( e->has_flag( "REACH_ATTACK" ) ) {
+            res.emplace( "REACH", item::gun_mode { e->tname(), const_cast<item *>( e ),
+                                                   e->has_flag( "REACH3" ) ? 3 : 2, true } );
+        }
+        if( e->is_gunmod() ) {
+            for( auto m : e->type->gunmod->mode_modifier ) {
+                res.emplace( m.first, item::gun_mode { m.second.first, const_cast<item *>( this ),
+                                                       m.second.second, false } );
+            }
+        }
+    }
+
+    return res;
+}
+
+const item::gun_mode item::gun_has_mode( const std::string& mode ) const
+{
+    if( is_gun() ) {
+        for( auto e : gun_all_modes() ) {
+            if( e.first == mode ) {
+                return e.second;
+            }
+        }
+    }
+    return { "", nullptr, 0, false };
```

Comment:
You could yank it out into a `static const`, then you could have the function return a reference.